src/conn/conn_compact.c

```diff
@@ -0,0 +1,196 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+
+/*
+ * __compact_server_run_chk --
+ *     Check to decide if the compact server should continue running.
+ */
+static bool
+__compact_server_run_chk(WT_SESSION_IMPL *session)
+{
+    return (FLD_ISSET(S2C(session)->server_flags, WT_CONN_SERVER_COMPACT));
+}
+
+/*
+ * __compact_server --
+ *     The compact server thread.
+ */
+static WT_THREAD_RET
+__compact_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_CURSOR *cursor;
+    WT_DECL_RET;
+    WT_SESSION *wt_session;
+    WT_SESSION_IMPL *session;
+    int exact;
+    const char *key;
+    bool full_iteration;
+
+    session = arg;
+    conn = S2C(session);
+    wt_session = (WT_SESSION *)session;
+    cursor = NULL;
+    key = NULL;
+    exact = 0;
+    full_iteration = false;
+
+    for (;;) {
+
+        /* When the entire metadata file has been parsed, take a break. */
+        if (full_iteration) {
+            full_iteration = false;
+            /*
+             * TODO: We may have processed just a few files, might as well sleep to give the system
+             * a break.
+             */
+            __wt_sleep(10, 0);
```

Comment:
You can add time to wait for the condition variable when the complete iteration is finished so that it can be stopped if we want during that time instead of sleep. Use 0 timestamps if the compact server is disabled.

Reply:
See this change: 8b302f31816200fb7c26cc9731bb40b15dbf33df. I added a timeout so the logic had to be updated a tiny bit. Let me know what you think.